File: src/core/render.ts
// Shared deterministic text renderers used by status, waves, report, context, and check.
import { criticalPath, groupWaves, nextRunnable, type DagTask } from "./dag.js";
import type { State, TaskStatus } from "./state.js";
import { readArtifact } from "./specFiles.js";
import { stripHtmlComments } from "./md.js";

export const GLYPH: Record<TaskStatus, string> = {
  complete: "✓",
  running: "◐",
  pending: "○",
  blocked: "⚠",
};

export interface Counts {
  pending: number;
  running: number;
  complete: number;
  blocked: number;
  total: number;
}

export function counts(state: State): Counts {
  const c: Counts = { pending: 0, running: 0, complete: 0, blocked: 0, total: 0 };
  for (const t of Object.values(state.tasks)) {
    c[t.status]++;
    c.total++;
  }
  return c;
}

export function dagTasks(state: State): DagTask[] {
  return Object.values(state.tasks).map((t) => ({
    id: t.id,
    wave: t.wave,
    depends: t.depends,
    status: t.status,
  }));
}

/** Render the §7.3.3 wave graph: tasks grouped by wave with status glyphs + critical path. */
export function waveGraph(state: State): string {
  const tasks = dagTasks(state);
  if (tasks.length === 0) return "(no tasks yet)";
  const lines: string[] = [];
  for (const row of groupWaves(tasks)) {
    lines.push(`Wave ${row.wave}`);
    for (const t of row.tasks) {
      const full = state.tasks[t.id];
      let line = `  ${GLYPH[t.status]} ${t.id}  ${full.title}`;
      if (t.status === "blocked" && full.blocker) line += `  (blocked: ${full.blocker})`;
      lines.push(line);
    }
  }
  const cp = criticalPath(tasks);
  if (cp.length) lines.push("", `Critical path: ${cp.join(" → ")}`);
  return lines.join("\n");
}

/** One-line human summary of next-runnable result. */
export function nextSummary(state: State): string {
  const r = nextRunnable(dagTasks(state));
  switch (r.kind) {
    case "task": return `${r.id} — ${state.tasks[r.id].title}`;
    case "all-complete": return "all tasks complete";
    case "all-blocked": return `all remaining blocked: ${r.blocked.join(", ")}`;
    case "waiting": return `waiting on: ${r.blocking.join(", ")}`;
  }
}

// --- Live-signal renderers for `specd context` (G3) --------------------------------------------

/** Blocker one-liners, e.g. "T4: api key missing". Empty array if none recorded. */
export function blockerLines(state: State): string[] {
  return state.blockers.map((b) => `${b.task}: ${b.reason}`);
}

export interface MidreqSummary { turn: number; impact: string; input: string; }

/**
 * Parse the most recent `## Turn N — … — impact: X` block from mid-requirements.md so `context`
 * can show *what* changed when a midreq raised the awaiting-approval gate. Returns null if the
 * ledger is absent or has no parseable entry. Deterministic string parse — no deps.
 */
export function latestMidreq(root: string, slug: string): MidreqSummary | null {
  const raw = readArtifact(root, slug, "mid-requirements.md");
  if (!raw) return null;
  const idx = raw.lastIndexOf("## Turn ");
  if (idx === -1) return null;
  const block = raw.slice(idx);
  const header = block.match(/^##\s+Turn\s+(\d+).*?impact:\s*(\w+)/i);
  if (!header) return null;
  const im = block.match(/\*\*User input \(verbatim\):\*\*\s*"?(.*?)"?\s*$/im);
  let input = im ? im[1].trim() : "";
  if (input.length > 120) input = input.slice(0, 117) + "...";
  return { turn: parseInt(header[1], 10), impact: header[2].toLowerCase(), input };
}

/** Requirement numbers declared in requirements.md (`## Requirement N`). Shared with check gate 7. */
export function requirementNumbers(reqMd: string): Set<number> {
  return new Set(
    [...stripHtmlComments(reqMd).matchAll(/^##\s+Requirement\s+(\d+)/gim)].map((m) => parseInt(m[1], 10)),
  );
}

/**
 * Requirement numbers defined in requirements.md but referenced by zero tasks (state-side
 * coverage). Sorted ascending. Used by `context` at the VERIFY gate to list what still needs
 * confirming. Empty if reqMd is absent or every requirement is covered.
 */
export function uncoveredRequirements(state: State, reqMd: string | null): number[] {
  if (!reqMd) return [];
  const referenced = new Set<number>();
  for (const t of Object.values(state.tasks)) for (const n of t.requirements) referenced.add(n);
  return [...requirementNumbers(reqMd)].filter((n) => !referenced.has(n)).sort((a, b) => a - b);
}

/**
 * Evaluate the G5 acceptance ledger against the requirements: `unmet` is every defined requirement
 * with no passing criterion, `failed` is every criterion key recorded as `fail`. Both sorted.
 * Shared by `approve` (the VERIFY refusal when `gates.acceptance: required`) and any reporting.
 */
export function acceptanceGaps(state: State, reqMd: string | null): { unmet: number[]; failed: string[] } {
  const reqs = reqMd ? [...requirementNumbers(reqMd)] : [];
  const map = state.acceptance ?? {};
  const passed = new Set<number>();
  const failed: string[] = [];
  for (const [key, rec] of Object.entries(map)) {
    if (rec.status === "pass") passed.add(rec.requirement);
    else failed.push(key);
  }
  return {
    unmet: reqs.filter((n) => !passed.has(n)).sort((a, b) => a - b),
    failed: failed.sort((a, b) => a.localeCompare(b, undefined, { numeric: true })),
  };
}

File: internal/core/render.go
package core

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// Glyph maps each TaskStatus to the single-character symbol used when
// rendering task lists and wave graphs.
var Glyph = map[TaskStatus]string{
	TaskComplete: "✓",
	TaskRunning:  "◐",
	TaskPending:  "○",
	TaskBlocked:  "⚠",
}

// Counts tallies tasks by status (pending, running, complete, blocked) plus
// the overall total, as reported by CountTasks.
type Counts struct {
	Pending  int `json:"pending"`
	Running  int `json:"running"`
	Complete int `json:"complete"`
	Blocked  int `json:"blocked"`
	Total    int `json:"total"`
}

// CountTasks tallies the tasks in state by their Status into a Counts
// summary.
func CountTasks(state *State) Counts {
	var c Counts
	for _, t := range state.Tasks {
		switch t.Status {
		case TaskPending:
			c.Pending++
		case TaskRunning:
			c.Running++
		case TaskComplete:
			c.Complete++
		case TaskBlocked:
			c.Blocked++
		}
		c.Total++
	}
	return c
}

// DagTasksFromState converts state's tasks into the []DagTask form used by
// the DAG/critical-path algorithms (WaveGraph, NextRunnable, CriticalPath).
func DagTasksFromState(state *State) []DagTask {
	out := make([]DagTask, 0, len(state.Tasks))
	for _, t := range state.Tasks {
		out = append(out, DagTask{
			ID:      t.ID,
			Wave:    t.Wave,
			Depends: t.Depends,
			Status:  t.Status,
		})
	}
	return out
}

// WaveGraph renders a human-readable, wave-by-wave listing of state's
// tasks, including status glyphs, blocked reasons, and the critical path.
func WaveGraph(state *State) string {
	tasks := DagTasksFromState(state)
	if len(tasks) == 0 {
		return "(no tasks yet)"
	}
	var lines []string
	for _, row := range GroupWaves(tasks) {
		lines = append(lines, fmt.Sprintf("Wave %d", row.Wave))
		for _, t := range row.Tasks {
			full := state.Tasks[t.ID]
			line := fmt.Sprintf("  %s %s  %s", Glyph[t.Status], t.ID, full.Title)
			if t.Status == TaskBlocked && full.Blocker != nil {
				line += fmt.Sprintf("  (blocked: %s)", *full.Blocker)
			}
			lines = append(lines, line)
		}
	}
	cp := CriticalPath(tasks)
	if len(cp) > 0 {
		lines = append(lines, "", "Critical path: "+strings.Join(cp, " → "))
	}
	return strings.Join(lines, "\n")
}

// NextSummary returns a short, human-readable description of what to do
// next: the next runnable task, an all-complete message, an all-blocked
// message, or a waiting-on message.
func NextSummary(state *State) string {
	r := NextRunnable(DagTasksFromState(state))
	switch r.Kind {
	case NextTask:
		t := state.Tasks[r.ID]
		return fmt.Sprintf("%s — %s", r.ID, t.Title)
	case NextAllComplete:
		return "all tasks complete"
	case NextAllBlocked:
		return "all remaining blocked: " + strings.Join(r.Blocked, ", ")
	case NextWaiting:
		return "waiting on: " + strings.Join(r.Blocking, ", ")
	}
	return ""
}

// BlockerLines renders state's recorded blockers as "<task>: <reason>"
// lines, one per blocker.
func BlockerLines(state *State) []string {
	out := make([]string, len(state.Blockers))
	for i, b := range state.Blockers {
		out[i] = fmt.Sprintf("%s: %s", b.Task, b.Reason)
	}
	return out
}

// MidreqSummary is a condensed view of the most recent mid-spec
// requirement-change turn: its turn number, impact level, and the verbatim
// user input that triggered it.
type MidreqSummary struct {
	Turn   int
	Impact string
	Input  string
}

var (
	midreqHeaderRe = regexp.MustCompile(`(?i)^##\s+Turn\s+(\d+).*?impact:\s*(\w+)`)
	midreqInputRe  = regexp.MustCompile(`(?im)\*\*User input \(verbatim\):\*\*\s*"?(.*?)"?\s*$`)
)

// LatestMidreq reads root/slug's mid-requirements.md artifact and returns a
// summary of the most recent turn, or nil if the artifact is missing or has
// no parseable turn.
func LatestMidreq(root, slug string) *MidreqSummary {
	raw := ReadArtifact(root, slug, "mid-requirements.md")
	if raw == nil {
		return nil
	}
	idx := strings.LastIndex(*raw, "## Turn ")
	if idx == -1 {
		return nil
	}
	block := (*raw)[idx:]
	headerM := midreqHeaderRe.FindStringSubmatch(block)
	if headerM == nil {
		return nil
	}
	turn := 0
	fmt.Sscanf(headerM[1], "%d", &turn)
	impact := strings.ToLower(headerM[2])
	inputM := midreqInputRe.FindStringSubmatch(block)
	input := ""
	if inputM != nil {
		input = strings.TrimSpace(inputM[1])
	}
	if len(input) > 120 {
		input = input[:117] + "..."
	}
	return &MidreqSummary{Turn: turn, Impact: impact, Input: input}
}

var reqNumRe = regexp.MustCompile(`(?im)^##\s+Requirement\s+(\d+)`)

// RequirementNumbers extracts the set of requirement numbers declared by
// `## Requirement N` headers in reqMd, after stripping HTML comments.
func RequirementNumbers(reqMd string) map[int]bool {
	m := reqNumRe.FindAllStringSubmatch(StripHTMLComments(reqMd), -1)
	out := make(map[int]bool, len(m))
	for _, match := range m {
		var n int
		fmt.Sscanf(match[1], "%d", &n)
		out[n] = true
	}
	return out
}

// UncoveredRequirements returns the requirement numbers declared in reqMd
// that no task in state references, sorted ascending, or nil if reqMd is
// nil.
func UncoveredRequirements(state *State, reqMd *string) []int {
	if reqMd == nil {
		return nil
	}
	referenced := make(map[int]bool)
	for _, t := range state.Tasks {
		for _, n := range t.Requirements {
			referenced[n] = true
		}
	}
	reqNums := RequirementNumbers(*reqMd)
	var out []int
	for n := range reqNums {
		if !referenced[n] {
			out = append(out, n)
		}
	}
	sort.Ints(out)
	return out
}

// AcceptanceGaps reports the acceptance-criteria gaps for a spec:
// requirement numbers with no passing acceptance record, and the keys of
// acceptance records that failed.
type AcceptanceGaps struct {
	Unmet  []int    `json:"unmet"`
	Failed []string `json:"failed"`
}

// GetAcceptanceGaps computes the AcceptanceGaps for state against reqMd:
// requirement numbers with no passing acceptance record, and the keys of
// any failed acceptance records.
func GetAcceptanceGaps(state *State, reqMd *string) AcceptanceGaps {
	var reqs []int
	if reqMd != nil {
		nums := RequirementNumbers(*reqMd)
		for n := range nums {
			reqs = append(reqs, n)
		}
	}
	passed := make(map[int]bool)
	var failed []string
	for key, rec := range state.Acceptance {
		if rec.Status == "pass" {
			passed[rec.Requirement] = true
		} else {
			failed = append(failed, key)
		}
	}
	var unmet []int
	for _, n := range reqs {
		if !passed[n] {
			unmet = append(unmet, n)
		}
	}
	sort.Ints(unmet)
	sort.Strings(failed)
	return AcceptanceGaps{Unmet: unmet, Failed: failed}
}
